### ISSUE TITLE
CEC Bit timing configuration?

### ISSUE THREAD

**Issue by drsamko** [Pulse-Eight/libcec#566]
Hi there,

we're using LibCEC on a Pi 4 to verify our AVR for certification. One requirement is to test the support for the full range of bit timing (e.g. Logical '1' low level between .4 and .8ms etc.),

Currently the controller is using the standard values (i.e. .6ms for the example above). Is there a way to configure the CEC driver and update the bit timing? I'm particularly looking into sending an <Abort> message with Logical '1' Low timing length of 400us, as well as another test with Logical '1' High timing length of 1450us.

Any help is immensely appreciated!

**Comment by malard** [comment]
I don’t think you can do this because the timing is controlled at a SoC level on the Pi, however in theory this could be done on our external adapters but would need custom firmware to achieve this.

From: drsamko ***@***.***>
Sent: 06 May 2021 20:16
To: Pulse-Eight/libcec ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [Pulse-Eight/libcec] CEC Bit timing configuration? (#566)


Hi there,

we're using LibCEC on a Pi 4 to verify our AVR for certification. One requirement is to test the support for the full range of bit timing (e.g. Logical '1' low level between .4 and .8ms etc.),

Currently the controller is using the standard values (i.e. .6ms for the example above). Is there a way to configure the CEC driver and update the bit timing? I'm particularly looking into sending an message with Logical '1' Low timing length of 400us, as well as another test with Logical '1' High timing length of 1450us.

Any help is immensely appreciated!

**Comment by drsamko** [comment]
Thanks for the quick response, @malard, are there any public resources regarding updating the firmware on the external device (can we do it in-house)? We could potentially look into that if it can't be done on the Pi.

**Comment by malard** [comment]
Please email ***@***.*** as it would be a commercial effort

Thanks

Sent from my Huawei phone


-------- Original message --------
From: drsamko ***@***.***>
Date: Thu, 6 May 2021, 20:27
To: Pulse-Eight/libcec ***@***.***>
Cc: Martin Ellis ***@***.***>, Mention ***@***.***>
Subject: Re: [Pulse-Eight/libcec] CEC Bit timing configuration? (#566)

Thanks for the quick response, @malard<https://github.com/malard>, are there any public resources regarding updating the firmware on the external device (can we do it in-house)? We could potentially look into that if it can't be done on the Pi.

-

**Comment by drsamko** [comment]
Thanks, though your email is masked in your message. can you separate it please? We might possibly approach
If it's your email, I can grab it from your profile

**Comment by malard** [comment]
Sales at pulse-eight dot com

Sent from my Huawei phone


-------- Original message --------
From: drsamko ***@***.***>
Date: Thu, 6 May 2021, 20:32
To: Pulse-Eight/libcec ***@***.***>
Cc: Martin Ellis ***@***.***>, Mention ***@***.***>
Subject: Re: [Pulse-Eight/libcec] CEC Bit timing configuration? (#566)

Thanks, though your email is masked in your message. can you separate it please? We might possibly approach
If it's your email, I can grab it from your profile

-

**Comment by popcornmix** [comment]
If you use the kms driver on Pi and build libcec for linux (`-DHAVE_LINUX_API=1 -DHAVE_RPI_API=0`)
then the driver is available to change. I suspect you want to start looking here:
https://github.com/raspberrypi/linux/blob/53a5ac4935c500d32bfc465551cc5107e091c09c/drivers/gpu/drm/vc4/vc4_hdmi.c#L2270

**Comment by drsamko** [comment]
Thanks @popcornmix , looks promising!
